### PR TITLE
[SPARK-8348][SQL] Add in operator to DataFrame Column

### PR DESCRIPTION
**Pull request by yu-iskw**
I have added it for only Scala.

TODO: we should also support `in` operator in Python.

### CODE REVIEW

**Comment by shivaram** [comment]
As mentioned in the JIRA, we should also do this for the R API

**Comment by yu-iskw** [comment]
@shivaram Got it. Is there any documentation how to develop SparkR?

**Comment by shivaram** [comment]
We don't have a developer guide yet for SparkR, but it design is pretty similar to the PySpark. For example you can see this PR which added math operations to SparkR https://github.com/apache/spark/commit/50da9e89161faa0ecdc1feb3ffee6c822a742034

**Comment by rxin** [comment]
I think it is ok to add the R stuff later. It is pretty hard for everyone to learn R and add all the expressions. Also considering there is a selectExpr workaround ...

File: sql/core/src/test/scala/org/apache/spark/sql/ColumnExpressionSuite.scala [hunk 1]
```diff
@@ -296,6 +296,15 @@ class ColumnExpressionSuite extends QueryTest {
     checkAnswer(testData.filter($"a".between($"b", $"c")), expectAnswer)
   }
 
+  test("in") {
+    var expectedAnswer = testData2.collect().toSeq.filter(r => r.getInt(0) == 1 || r.getInt(0) == 2)
+    checkAnswer(testData2.filter($"a" in(1, 2)), expectedAnswer)
```
Comment (rxin) [inline sql/core/src/test/scala/org/apache/spark/sql/ColumnExpressionSuite.scala]: I think we prefer to have explicitly defined test data in the test cases. The testData/testData1 etc are from old test cases.

Take a look at test case "self join explode" for an example.

**Comment by yu-iskw** [comment]
@rxin Thank you for your feedback. I'll fix the test cases soon. 

As you said, it is hard to implement `%in%` operator in SparkR. And we don't have any operator for column and array/list. If you have any idea, could you please tell me that?
https://issues.apache.org/jira/browse/SPARK-8348?focusedCommentId=14590880&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-14590880

**Comment by yu-iskw** [comment]
Jenkins, test this please.

**Comment by yu-iskw** [comment]
@rxin Could you please check it? I have fixed the test cases. Thanks.

File: sql/core/src/test/scala/org/apache/spark/sql/ColumnExpressionSuite.scala [hunk 2]
```diff
@@ -296,6 +296,28 @@ class ColumnExpressionSuite extends QueryTest {
     checkAnswer(testData.filter($"a".between($"b", $"c")), expectAnswer)
   }
 
+  test("in") {
+    val df = Seq((1, "x"), (2, "y"), (3, "z")).toDF("a", "b")
+    checkAnswer(df.filter($"a" in(1, 2)),
```
Comment (rxin) [inline sql/core/src/test/scala/org/apache/spark/sql/ColumnExpressionSuite.scala]: can you not use infix notation here?

i.e. do
```scala
$"a".in(1, 2)
```

Comment (yu-iskw) [inline sql/core/src/test/scala/org/apache/spark/sql/ColumnExpressionSuite.scala]: I got it. I will modify those. Thanks.

**Comment by rxin** [comment]
Thanks. Merging.

**Comment by yu-iskw** [comment]
Thank you for merging it!

**Comment by rxin** [comment]
Can you add the Python version too?

**Comment by yu-iskw** [comment]
Yes, will do.

**Comment by yu-iskw** [comment]
I also created an issue for the python version.
https://issues.apache.org/jira/browse/SPARK-8477

File: sql/core/src/main/scala/org/apache/spark/sql/Column.scala [hunk 3]
```diff
@@ -621,7 +621,7 @@ class Column(protected[sql] val expr: Expression) extends Logging {
    * @since 1.3.0
    */
   @scala.annotation.varargs
-  def in(list: Column*): Column = In(expr, list.map(_.expr))
+  def in(list: Any*): Column = In(expr, list.map(lit(_).expr))
```
Comment (marmbrus) [inline sql/core/src/main/scala/org/apache/spark/sql/Column.scala]: We probably want to check if the thing passed in is a column, and avoid wrapping it in a literal if so.

Comment (rxin) [inline sql/core/src/main/scala/org/apache/spark/sql/Column.scala]: lit actually handles that itself (it just returns the object itself it is a column)

**Pull request closed without merging**